mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,754 @@
+
+                    // If we don't get expected type_url, then continue.
+                    // (This check exists for not allocating ParseTxDetailsArgs when we don't use
+                    // it. We could use it in match statement but that way we have to duplicate ~10
+                    // lines of code for each pattern.)
+                    if ![SEND_TYPE_URL, CLAIM_HTLC_TYPE_URL, CREATE_HTLC_TYPE_URL].contains(&msg.type_url.as_str()) {
+                        continue;
+                    }
+
+                    let args = ParseTxDetailsArgs {
+                        coin,
+                        storage,
+                        active_assets,
+                        internal_id,
+                        block_height: tx.height.into(),
+                        tx_hash: tx.hash.to_string(),
+                        tx_body_msg: msg.value.as_slice(),
+                        fee_details,
+                    };
+
+                    let details = match msg.type_url.as_str() {
+                        SEND_TYPE_URL => match parse_send_tx_details(args).await {
+                            Ok(details) => details,
+                            Err(e) => {
+                                return Err(Stopped {
+                                    phantom: Default::default(),
+                                    stop_reason: e,
+                                });
+                            },
+                        },
+                        CLAIM_HTLC_TYPE_URL => match parse_claim_htlc_tx_details(args).await {
+                            Ok(details) => details,
+                            Err(e) => {
+                                return Err(Stopped {
+                                    phantom: Default::default(),
+                                    stop_reason: e,
+                                });
+                            },
+                        },
+                        CREATE_HTLC_TYPE_URL => match parse_create_htlc_tx_details(args).await {
+                            Ok(details) => details,
+                            Err(e) => {
+                                return Err(Stopped {
+                                    phantom: Default::default(),
+                                    stop_reason: e,
+                                });
+                            },
+                        },
+                        _ => continue,
+                    };
+
+                    let details = match details {
+                        Some(details) => details,
+                        None => continue,
+                    };
+
+                    upsert_tx_details(&mut tx_details, details);
+                }
+
+                for (ticker, txs) in tx_details.into_iter() {
+                    let id = WalletId::new(ticker.replace('-', "_"));
+
+                    try_or_return_stopped_as_err!(
+                        storage
+                            .add_transactions_to_history(&id, txs)
```

Comment:
In this implementation we iterate over the `tx_details` map twice, and also store the whole TX history in memory.
Is it possible to call `storage.add_transactions_to_history` right after the TX is parsed?

Reply:
> and also keep the whole TX history in memory.

Max possible transaction count can be in memory is 50. And that's maximum which might happen only at initialiation.